File: src/main/java/net/scriptgate/masonry/transition/ArcedTransition.java
package net.scriptgate.masonry.transition;

import net.scriptgate.common.Point;
import net.scriptgate.masonry.api.Transition;

import java.util.function.BiFunction;

public class ArcedTransition extends LinearTransition {

    public static final BiFunction<Point, Point, Transition> ARCED = (from, to) -> new ArcedTransition(from.x, from.y, to.x, to.y);

    public static final int ARC_HEIGHT = 10;

    public ArcedTransition(Point from, Point to) {
        this(from.x, from.y, to.x, to.y);
    }

    public ArcedTransition(int xFrom, int yFrom, int xTo, int yTo) {
        super(xFrom, yFrom, xTo, yTo);
    }

    @Override
    public Point getLocation() {
        double offset = Math.sin(getPercentage() * Math.PI) * ARC_HEIGHT;

        //leans right when direction is down
        //leans left when direction is up
        //leans top when direction is left
        //leans bottom when direction is right
        double offsetDirection = direction - (Math.PI / 2f);

        int deltaX = (int) (offset * Math.cos(offsetDirection));
        int deltaY = (int) (offset * Math.sin(offsetDirection));

        double percentage = getPercentage() * distance;

        deltaX += (int) (percentage * Math.cos(direction));
        deltaY += (int) (percentage * Math.sin(direction));

        return new Point(pointFrom.x + deltaX, pointFrom.y + deltaY);
    }
}

File: src/main/kotlin/net/scriptgate/masonry/transition/ArcedTransition.kt
package net.scriptgate.masonry.transition

import net.scriptgate.common.Point
import net.scriptgate.masonry.api.Transition

import java.util.function.BiFunction

class ArcedTransition(xFrom: Int, yFrom: Int, xTo: Int, yTo: Int) : LinearTransition(xFrom, yFrom, xTo, yTo) {
    override fun getLocation(): Point {

        val offset = Math.sin(super.percentage * Math.PI) * ARC_HEIGHT

        //leans right when direction is down
        //leans left when direction is up
        //leans top when direction is left
        //leans bottom when direction is right
        val offsetDirection = direction - Math.PI / 2f

        var deltaX = (offset * Math.cos(offsetDirection)).toInt()
        var deltaY = (offset * Math.sin(offsetDirection)).toInt()

        val percentage = super.percentage * distance

        deltaX += (percentage * Math.cos(direction)).toInt()
        deltaY += (percentage * Math.sin(direction)).toInt()

        return Point(pointFrom.x + deltaX, pointFrom.y + deltaY)
    }


    constructor(from: Point, to: Point) : this(from.x, from.y, to.x, to.y)


    object ARCED : BiFunction<Point, Point, Transition> {
        override fun apply(from: Point, to: Point): Transition {
            return ArcedTransition(from.x, from.y, to.x, to.y)
        }
    }

    val ARC_HEIGHT = 10

}
